### PR TITLE
Koa support

### LINKED ISSUE #131: koa support
[koa](http://koajs.com/) is a web framework taking many ideas from Express, but ditching callbacks in favour of generators. Is native support of koa in endpoints planned? Which part of code should I look into if I want to do it myself?

### PR DESCRIPTION
**Pull request by Dremora**
Close #131.
A few notes:
- Framework-specific functionality has been extracted into separate modules — I simply called them frameworks, but there might be a better name (framework adapters? response handlers? (but this name is already used)).
- Instead of implementing framework detection, I've added a new parameter to controllers called `framework`.
- Each framework accepts two arguments. 
  - `buildPayload`, a function which takes request object and returns a promise wrapping an object returned by payload handler. This allows us to define logic for framework-specific request handling and response forming to the framework handler, and leave the rest to other components. This could have been slightly simpler, but is done this way for koa: all asynchronous code should be wrapped in a promise executed with `yield`.
  - `responder`, an optional argument I simply migrated from existing code. Not sure if it makes any sense to keep it.
- Return value is the framework-specific handler: plain function in case of express and hapi and a generator function in case of koa.
- I've blacklisted ES6 generators in babel arguments since we want to leave them intact — koa runs only on Node 0.12 where generators are natively supported.

### CODE REVIEW

**Comment by tkellen** [comment]
This is great @Dremora, thank you so much! I'm really excited to see someone using endpoints with another http server.

So, we don't need to pass the responder method around, what you've built *is* the responder. What do you think about just re-using that name? We'd have `responder-hapi`, `responder-express`, `responder-koa`.

Some things to address:
* [ ] Add JSDoc comments for all functions added
* [ ] The ES5 rebuild should be in a commit by itself so the diff is less painful to read.
* [ ] Add a migration guide--we can release this as `0.10` soon.
* [ ] Update example repo (e.g. put all of the current code in an express folder and `cp -R` it to a koa folder an update what's needed for both).

File: src/controller/lib/handle.js [hunk 1]
```diff
@@ -12,23 +14,23 @@ module.exports = function (config, baseUrl) {
...
-        .then(format)
-        .then(respond)
-        .catch(function (err) {
-          //throw err;
-          return respond(payloadHandler.error(err));
-        });
-    }
+  var buildPayload = function (request) {
+    var errors = validate(request);
+    return bPromise[errors ? 'reject' : 'resolve'](errors)
```
Comment (tkellen) [inline src/controller/lib/handle.js]: I'd like to be able to unit test this directly. Can we move it into the `lib` folder and add the arguments needed so we can partially apply it here?

File: src/controller/lib/handle.js [hunk 2]
```diff
@@ -12,23 +14,23 @@ module.exports = function (config, baseUrl) {
...
-    } else {
-      process(request)
-        .then(format)
-        .then(respond)
-        .catch(function (err) {
-          //throw err;
-          return respond(payloadHandler.error(err));
-        });
-    }
+  var buildPayload = function (request) {
```
Comment (tkellen) [inline src/controller/lib/handle.js]: I'd like to be able to unit test `buildPayload` directly. Can we move it into the `lib` folder and add the arguments needed so we can partially apply it here?

**Comment by Dremora** [comment]
Responder sounds like a perfect name! I'll make all the required changes.

File: src/controller/lib/handle.js (hunk 2 above)
Comment (Dremora) [inline src/controller/lib/handle.js]: Sure, sounds good.

**Comment by tkellen** [comment]
Just FYI, @bobholt and I have the architecture we're aiming for laid out here:
https://github.com/endpoints/endpoints/wiki/Architecture-Views-DecompositionUses

It has been some time since he and I discussed this, but I believe what we have here is a partial implementation of [ResHandler](https://github.com/endpoints/endpoints/wiki/Architecture-Views-DecompositionUses#http-endpoints-reqhandler). The responder method may end up being a part of that eventually.

**Comment by tkellen** [comment]
@Dremora Any updates here? I'm super excited to land your changes.

**Comment by ashleygwilliams** [comment]
:+1: to @tkellen's comment

**Comment by Dremora** [comment]
@tkellen sorry, was super busy lately — I'll make the changes now.

**Comment by tkellen** [comment]
Thanks @Dremora! I'm going to be working on cutting a release this weekend and this will be included!

**Comment by Dremora** [comment]
Sorry for not completing the docs/example — I've been even more busy during the last month juggling between two jobs. The code should be fine though, I've followed all the comments.

**Comment by tkellen** [comment]
No worries @Dremora, I know how it goes. I'd still really love to see an example repo!

**Pull request merged**